system/Router/RouteCollection.php

```diff
@@ -263,8 +279,26 @@ public function loadRoutes(string $routesFile = APPPATH . 'Config/Routes.php')
             return $this;
         }
 
+        // Include the passed in routesFile if it doesn't exist.
+        // Only keeping that around for BC purposes for now.
+        $routeFiles = $this->routeFiles;
+        if (! in_array($routesFile, $routeFiles, true)) {
+            $routeFiles[] = $routesFile;
+        }
+
+        // We need this var in local scope
+        // so route files can access it.
         $routes = $this;
-        require $routesFile;
+
+        foreach ($routeFiles as $routesFile) {
+            if (! is_file($routesFile)) {
+                log_message('warning', 'Routes file not found: ' . $routesFile . '.');
```

Comment:
probably sprintf is cleaner:

```suggestion
                log_message('warning', sprintf('Routes file not found: %s.', $routesFile));
```

Reply:
Thank you for reveiw!

By the way, is the last dot necessary?

Reply:
last dot is same with the previous usage ` . $routesFile . '.'` to keep same value.

Reply:
I mean, isn't `log_message('warning', 'Routes file not found: ' . $routesFile);` sufficient?

Reply:
I have no opinion on that ,probably @MGatner can give opinion on it

Reply:
```
Cast.php
    'invalidCastMethod'      => 'The "{0}" is invalid cast method, valid methods are: ["get", "set"].',
CLI.php
        'fileCreate'       => 'File created: {0}',
        'fileError'        => 'Error while creating file: "{0}"',
        'fileExist'        => 'File exists: "{0}"',
        'fileOverwrite'    => 'File overwritten: "{0}"',
        'usingCINamespace' => 'Warning: Using the "CodeIgniter" namespace will generate the file in the system directory.',
    'invalidColor'        => 'Invalid "{0}" color: "{1}".',
Cookie.php
    'invalidExpiresTime'    => 'Invalid "{0}" type for "Expires" attribute. Expected: string, integer, DateTimeInterface object.',
Core.php
    'invalidFile'                  => 'Invalid file: "{0}"',
    'invalidPhpVersion'            => 'Your PHP version must be {0} or higher to run CodeIgniter. Current version: {1}',
    'missingExtension'             => 'The framework needs the following extension(s) installed and loaded: "{0}".',
Database.php
    'invalidAllowedFields'             => 'Allowed fields must be specified for model: "{0}"',
Email.php
    'invalidAddress'       => 'Invalid email address: "{0}"',
    'attachmentMissing'    => 'Unable to locate the following email attachment: "{0}"',
    'attachmentUnreadable' => 'Unable to open this attachment: "{0}"',
    'noRecipients'         => 'You must include recipients: To, Cc, or Bcc',
    'sent'                 => 'Your message has been successfully sent using the following protocol: {0}',
    'SMTPError'            => 'The following SMTP error was encountered: {0}',
    'noSMTPAuth'           => 'Error: You must assign an SMTP username and password.',
    'failedSMTPLogin'      => 'Failed to send AUTH LOGIN command. Error: {0}',
    'SMTPAuthUsername'     => 'Failed to authenticate username. Error: {0}',
    'SMTPAuthPassword'     => 'Failed to authenticate password. Error: {0}',
    'SMTPDataFailure'      => 'Unable to send data: {0}',
    'exitStatus'           => 'Exit status code: {0}',
Encryption.php
    'authenticationFailed' => 'Decrypting: authentication failed.',
Fabricator.php
    'createFailed'      => 'Fabricator failed to insert on table "{0}": {1}',
Files.php
    'fileNotFound'      => 'File not found: "{0}"',
    'cannotMove'        => 'Could not move file "{0}" to "{1}". Reason: {2}',
Format.php
    'invalidJSON'      => 'Failed to parse JSON string. Error: {0}',
    'invalidMime'      => 'No Formatter defined for mime type: "{0}".',
HTTP.php
    'sslCertNotFound' => 'SSL certificate not found at: "{0}"',
    'curlError'       => '{0} : {1}',
    'invalidNegotiationType' => '"{0}" is not a valid negotiation type. Must be one of: media, charset, encoding, language.',
    'invalidHTTPProtocol' => 'Invalid HTTP Protocol Version. Must be one of: {0}',
    'cannotSetFilepath'      => 'When setting binary cannot set filepath: "{0}"',
    'cannotSetStatusCode'    => 'It does not support change status code for downloading. code: {0}, reason: {1}',
    'unknownStatusCode'     => 'Unknown HTTP status code provided with no message: {0}',
    'cannotParseURI'       => 'Unable to parse URI: "{0}"',
    'segmentOutOfRange'    => 'Request URI segment is out of range: "{0}"',
    'invalidPort'          => 'Ports must be between 0 and 65535. Given: {0}',
    'controllerNotFound' => 'Controller or its method is not found: {0}::{1}',
    'methodNotFound'     => 'Controller method is not found: "{0}"',
    'localeNotSupported' => 'Locale is not supported: {0}',
    'moveFailed'   => 'Could not move file "{0}" to "{1}". Reason: {2}',
    'uploadErrNoTmpDir'  => 'File could not be uploaded: missing temporary directory.',
    'invalidSameSiteSetting' => 'The SameSite setting must be None, Lax, Strict, or a blank string. Given: {0}',
Images.php
    'invalidDirection'       => 'Flip direction can be only "vertical" or "horizontal". Given: "{0}"',
Migrations.php
    'notFound'      => 'Migration file not found: ',
    'batchNotFound' => 'Target batch not found: ',
    'gap'           => 'There is a gap in the migration sequence near version number: ',
    'rollingBack'       => 'Rolling back migrations to batch: ',
    'removed'           => 'Rolling back: ',
    'added'             => 'Running: ',
    'on'        => 'Migrated On: ',
Publisher.php
    'destinationNotAllowed' => 'Destination is not on the allowed list of Publisher directories: "{0}"',
    'fileNotAllowed'        => '"{0}" fails the following restriction for "{1}": {2}',
Router.php
    'invalidDynamicController' => 'A dynamic controller is not allowed for security reasons. Route handler: "{0}"',
    'invalidControllerName'    => 'The namespace delimiter is a backslash (\), not a slash (/). Route handler: "{0}"',
Security.php
    'invalidSameSite' => 'The SameSite value must be None, Lax, Strict, or a blank string. Given: "{0}"',
Session.php
    'invalidSavePath'        => 'Session: Configured save path "{0}" is not a directory, does not exist or cannot be created.',
    'writeProtectedSavePath' => 'Session: Configured save path "{0}" is not writable by the PHP process.',
    'emptySavePath'          => 'Session: No save path configured.',
    'invalidSavePathFormat'  => 'Session: Invalid Redis save path format: "{0}"',
    'invalidSameSiteSetting' => 'Session: The SameSite setting must be None, Lax, Strict, or a blank string. Given: "{0}"',
Time.php
    'invalidMonth'   => 'Months must be between 1 and 12. Given: {0}',
    'invalidDay'     => 'Days must be between 1 and 31. Given: {0}',
    'invalidOverDay' => 'Days must be between 1 and {0}. Given: {1}',
    'invalidHours'   => 'Hours must be between 0 and 23. Given: {0}',
    'invalidMinutes' => 'Minutes must be between 0 and 59. Given: {0}',
    'invalidSeconds' => 'Seconds must be between 0 and 59. Given: {0}',
Validation.php
    'alpha_numeric_punct'   => 'The {field} field may contain only alphanumeric characters, spaces, and  ~ ! # $ % & * - _ + = | : . characters.',
    'equals'                => 'The {field} field must be exactly: {param}.',
    'in_list'               => 'The {field} field must be one of: {param}.',
    'not_equals'            => 'The {field} field cannot be: {param}.',
    'not_in_list'           => 'The {field} field must not be one of: {param}.',
View.php
    'invalidCellClass'      => 'Unable to locate view cell class: "{0}".',
    'tagSyntaxError'        => 'You have a syntax error in your Parser tags: "{0}"',
```

Reply:
```
BaseConnection.php
            log_message('error', 'Error connecting to the database: ' . $e);
                        log_message('error', 'Error connecting to the database: ' . $e);
                        log_message('error', 'Database: Failure during an automated transaction commit/rollback!');
BasePreparedQuery.php
                        log_message('error', 'Database: Failure during an automated transaction commit/rollback!');
CacheFactory.php
            log_message('critical', $e . ' Resorting to using ' . $backup . ' handler.');
CLI.php
            log_message('error', (string) $e);
Common.php
    function log_message(string $level, string $message, array $context = [])
Connection.php
                    log_message('error', $message);
                    log_message('error', "Database: Unable to set the configured connection charset ('{$this->charset}').");
            log_message('error', (string) $e);
Connection.php
            log_message('error', (string) $e);
Connection.php
            log_message('error', (string) $e);
Connection.php
            log_message('error', $error['message']);
Connection.php
            log_message('error', (string) $e);
CookieStore.php
                log_message('error', (string) $e);
Email.php
            log_message('error', 'Email: ' . $method . ' throwed ' . $e);
            log_message('error', 'Email: ' . $message);
            log_message('error', $this->printDebuggerRaw());
                log_message('error', 'Email: ' . $method . ' throwed ' . $e);
Exceptions.php
            log_message('critical', "{message}\nin {exFile} on line {exLine}.\n{trace}", [
        log_message(
FileHandler.php
                log_message('debug', 'Failed to set mode on cache file: ' . $e);
GeneratorTrait.php
            log_message('error', (string) $e);
RedisHandler.php
                // Note:: I'm unsure if log_message() is necessary, however I'm not 100% comfortable removing it.
                log_message('error', 'Cache: Redis connection failed. Check your configuration.');
                log_message('error', 'Cache: Redis authentication failed.');
                log_message('error', 'Cache: Redis select database failed.');
RedisHandler.php
            log_message('error', 'Session: Unable to obtain lock for ' . $this->keyPrefix . $sessionID . ' after 30 attempts, aborting.');
            log_message('debug', 'Session: Lock for ' . $this->keyPrefix . $sessionID . ' had no TTL, overriding.');
ResponseTrait.php
            log_message('error', (string) $e);
Router.php
        log_message('info', 'Used the default controller.');
Security.php
        log_message('info', 'CSRF token verified.');
        log_message('info', 'CSRF cookie sent.');
Toolbar.php
                log_message(
```

Reply:
It seems we often use the pattern like `Invalid file: "{0}"`.

Reply:
Fixed. 64130a27b5bd6440f2dcfc30b44410f669651f51